### PR TITLE
Function style fixer

### LINKED ISSUE #32: Different bash function syntaxes are tolerated
Hi,
I noticed that beautysh is not enforcing any specific style for the function declarations:

A)

```
function myfunc()
{
    ...
}
```

is accepted as well as
B)

```
function myfunc
{
    ...
}
```

and as well as
C)

```
myfunc()
{
    ...
}
```

I think beautysh should stick with one and enforce it.
What do you think?

### PR DESCRIPTION
**Pull request by f18m**
This PR is the first attempt to fix #32

### CODE REVIEW

File: beautysh/beautysh.py [hunk 1]
```diff
@@ -5,6 +5,22 @@
 import re
 import sys
 
+# correct function style detection is obtained only if following regex are tested in sequence.
+# styles are listed as follows:
+# 0) function keyword, open/closed parentheses
```
Comment (bemeurer) [inline beautysh/beautysh.py]: I think these are more instructive if you show an example, like
`function foo()`
`function foo`
`foo()`

File: beautysh/beautysh.py [hunk 2]
```diff
@@ -160,6 +200,12 @@ def beautify_string(self, data, path=''):
                                 inc += 1
                                 choice_case = -1
 
+                        # detect functions
+                        func_decl_style = self.detect_function_style(test_record)
+                        if func_decl_style != None:
+                             #sys.stderr.write("Found function declaration on line %d [%s] with style %d\n" % (line, stripped_record, func_decl_style))
```
Comment (bemeurer) [inline beautysh/beautysh.py]: Should this still be here?

File: beautysh/beautysh.py [hunk 3]
```diff
@@ -224,18 +270,28 @@ def main(self):
                                  "any in-place beautify.")
         parser.add_argument('--tab', '-t', action='store_true',
                             help="Sets indentation to tabs instead of spaces")
+        parser.add_argument('--force-function-style', '-s', nargs=1, type=int, default=1,
```
Comment (bemeurer) [inline beautysh/beautysh.py]: I think I prefer we call them something, and then document it, than using ints as the values.
Something like, `fnpar`,`fnnopar`,`par` for `function foo()`, `function foo`, and `foo()` respectively.

What do you think?

File: beautysh/beautysh.py [hunk 4]
```diff
@@ -31,6 +48,44 @@ def write_file(self, fp, data):
         with open(fp, 'w') as f:
             f.write(data)
 
+    def detect_function_style(self, test_record):
```
Comment (bemeurer) [inline beautysh/beautysh.py]: Could you add some comments here? I have some concerns about how robust this function is.

File: beautysh/beautysh.py [hunk 5]
```diff
@@ -31,6 +48,44 @@ def write_file(self, fp, data):
...
+        if func_decl_style is None:
+            return stripped_record
+        if self.apply_function_style is None:
+            # user does not want to enforce any specific function style
+            return stripped_record
+        regex = FUNCTION_STYLE_REGEX[func_decl_style]
+        replacement = FUNCTION_STYLE_REPLACEMENT[self.apply_function_style]
+        return re.sub(regex, replacement, stripped_record)
+
+    def get_test_record(self, source_line):
```
Comment (bemeurer) [inline beautysh/beautysh.py]: Nice

File: beautysh/beautysh.py [hunk 6]
```diff
@@ -224,18 +270,28 @@ def main(self):
...
                             help="Sets indentation to tabs instead of spaces")
+        parser.add_argument('--force-function-style', '-s', nargs=1, type=int, default=1,
+                            help="Force a specific Bash function formatting")
         args = parser.parse_args()
         if (len(sys.argv) < 2):
             parser.print_help()
             exit()
         if(type(args.indent_size) is list):
             args.indent_size = args.indent_size[0]
+        if not args.files:
```
Comment (bemeurer) [inline beautysh/beautysh.py]: Nice

File: beautysh/beautysh.py (hunk 1 above)
Comment (f18m) [reply beautysh/beautysh.py]: you're right, added!

File: beautysh/beautysh.py (hunk 2 above)
Comment (f18m) [reply beautysh/beautysh.py]: removed :)

File: beautysh/beautysh.py (hunk 3 above)
Comment (bemeurer) [reply beautysh/beautysh.py]: Agreed, fnonly is better!

Comment (bemeurer) [reply beautysh/beautysh.py]: Actually, having `fnonly` seems a bit inconsistent... What about `fnpar`, `fn`, and `par` or `fnpar`, `fnonly`, and `paronly`?

File: beautysh/beautysh.py (hunk 4 above)
Comment (f18m) [reply beautysh/beautysh.py]: ok, done

File: beautysh/beautysh.py (hunk 3 above)
Comment (f18m) [reply beautysh/beautysh.py]: yeah, you're right. I just checked in the last set of words (fnpar, fnonly, and paronly)

**Comment by f18m** [comment]
> Looking good!

great!

> I think now, to implement 2) and detect the most used syntax for a default behavior we just have to count the regex matches per style in the whole file?

yes right. To enable such behavior perhaps we could have a
   --force-function-style=autodetect 
option value?

**Comment by bemeurer** [comment]
>yes right. To enable such behavior perhaps we could have a
--force-function-style=autodetect
option value?

Agreed, but I'll leave that for a separate PR. This looks great! Thanks for your contributions!

**Pull request merged**